packages/react-compose/src/next/getSlots.tsx

```diff
@@ -0,0 +1,52 @@
+import * as React from 'react';
+import { getNativeElementProps, omit } from '@uifabric/utilities';
+import { GenericDictionary } from './types';
+import { nullRender } from './nullRender';
+
+/**
+ * Given the state and an array of slot names, will break out `slots` and `slotProps`
+ * collections.
+ *
+ * The root is always derived from the `as` prop.
+ *
+ * Slots will render as null if they are rendered as primitives with undefined children.
+ *
+ * The slotProps will always omit the `as` prop within them, and for slots that are string
+ * primitives, the props will be filtered according the the slot type. For example, if the
+ * slot is rendered `as: 'a'`, the props will be filtered for acceptable anchor props.
+ *
+ * @param state - State including slot definitions
+ * @param slotNames - Name of which props are slots
+ * @returns An object containing the `slots` map and `slotProps` map.
+ */
+export const getSlots = (state: GenericDictionary, slotNames?: string[] | undefined) => {
+  const slots: GenericDictionary = {
+    root: state.as || nullRender,
+  };
+  const slotProps: GenericDictionary = {
+    root: typeof state.as === 'string' ? getNativeElementProps(state.as, state) : omit(state, { as: 1 }),
+  };
+
+  if (slotNames) {
+    for (const name of slotNames) {
+      const slotDefinition = state[name];
+      const { as: slotAs, children } = slotDefinition;
+      const isSlotPrimitive = typeof slotAs === 'string';
+
+      slots[name] = slotDefinition.children !== undefined || !isSlotPrimitive ? slotAs : nullRender;
+
+      if (slots[name] !== nullRender) {
+        slotProps[name] = isSlotPrimitive
+          ? getNativeElementProps(slotAs, slotDefinition)
+          : omit(slotDefinition, { as: 1 });
+      }
+
+      if (children === 'function') {
```

Comment:
```suggestion
      if (typeof children === 'function') {
```

Reply:
```tsx
console.log(
  getSlots(
    {
      image: {
        as: "img",
        src: "bar.jpg",
        children: (Component, props) => <Component {...props} />
      }
    },
    ["image"]
  )
);
```

![image](https://user-images.githubusercontent.com/14183168/89168796-ad19d100-d57d-11ea-898f-c26f72491881.png)

`as` & `src` will go to `React.Fragment`, this will produce a React warning

Reply:
I added a test for this just to make sure we're good!

```jsx
it('can use slot children functions to replace default slot rendering', () => {
    expect(getSlots({ as: 'div', icon: { as: Foo, href: 'bar', children: (C, p) => <C {...p} /> } }, ['icon'])).toEqual(
      {
        slots: { root: 'div', icon: React.Fragment },
        slotProps: { root: {}, icon: { children: <Foo href="bar" /> } },
      },
    );
  });
```